video/out/vulkan/utils.c

```diff
@@ -1,8 +1,92 @@
+#include <pthread.h>
+
 #include "video/out/placebo/utils.h"
 #include "utils.h"
 
+static pthread_mutex_t queue_lock = {0};
+static PFN_vkGetDeviceProcAddr vk_get_device_proc_addr = NULL;
+
+/*
+ * Functions that require explicit synchronisation when accessing VkQueue
+ * objects.
+ *
+ * Additional documented functions that we don't currently use:
+ * - vkQueueWaitIdle
+ * - vkQueueBindSparse
+ * - vkQueueSubmit2[KHR]
+ *
+ * vkDeviceWaitIdle requires synchronising access to all queues.
+ * 
+ * vkDestroyDevice also technically requires synchronising access to all queues
+ * but we never do that with any other Vulkan code running.
+ */
+static PFN_vkQueueSubmit vk_queue_submit = NULL;
+static PFN_vkQueuePresentKHR vk_queue_present_khr = NULL;
+static PFN_vkDeviceWaitIdle vk_device_wait_idle = NULL;
+
+static VkResult mpvk_queue_submit(VkQueue queue, uint32_t submitCount, const VkSubmitInfo *pSubmits, VkFence fence) {
+    pthread_mutex_lock(&queue_lock);
+    VkResult result = vk_queue_submit(queue, submitCount, pSubmits, fence);
+    pthread_mutex_unlock(&queue_lock);
+    return result;
+}
+
+static VkResult mpvk_queue_present_khr(VkQueue queue, const VkPresentInfoKHR *pPresentInfo) {
+    pthread_mutex_lock(&queue_lock);
+    VkResult result = vk_queue_present_khr(queue, pPresentInfo);
+    pthread_mutex_unlock(&queue_lock);
+    return result;
+}
+
+static VkResult mpvk_device_wait_idle(VkDevice device) {
+    pthread_mutex_lock(&queue_lock);
+    VkResult result = vk_device_wait_idle(device);
+    pthread_mutex_unlock(&queue_lock);
+    return result;
+}
+
+static PFN_vkVoidFunction mpvk_get_device_proc_addr(VkDevice device, const char *pName) {
+
+    void *result = vk_get_device_proc_addr(device, pName);
+
+    if (strcmp(pName, "vkQueueSubmit") == 0) {
+        if (!vk_queue_submit) {
+            vk_queue_submit = result;
```

Comment:
To clarify what I mean further below: this doesn't work. There can be multiple `VkDevice` per `VkInstance`. The second invocation would override the first invocation's `vk_queue_submit` pointer, and end up with humbug.

Reply:
Possible work-around: use `vkGetInstanceProcAddr` instead of `vk_device_get_proc_addr` for these functions. That way it goes through the loader's own dispatch layer. That's, of course, one function call less efficient, but I doubt anybody cares here.